### ISSUE TITLE
Differentiate security errors from business errors? 

### ISSUE THREAD

**Issue by fmeriaux** [softwaremill/tapir#2654]
Currently, and to my knowledge,

The error output is unique. This does not allow you to write a lightweight implementation where you have a security error type and a business error type.

Then indeed the definition of an endpoint allows to define an output error as an Either, which allows to match each case to a valid output. 

```scala
endpoint
      .errorOut(
        oneOf[Either[SecurityError, BusinessError]](
          oneOfVariantFromMatchType(StatusCode.NotFound, emptyOutputAs(Right(BusinessEntityNotFound))),
          oneOfVariantFromMatchType(StatusCode.Unauthorized, emptyOutputAs(Left(Unauthorized))),
          oneOfVariantFromMatchType(StatusCode.Forbidden, emptyOutputAs(Left(Forbidden)))
        )
      )

```

But this approach has heavy consequences in the definition of the logic. It forces the safety logic to be composed with a function that encapsulates the standard Either in a new level of Either, and the same operation is necessary in the server logic.

```scala
   myEndpoint
     .serverSecurityLogic(SecurityPolicy.doStuff.andThen(_.map(_.left.map[Either[SecurityError, BusinessError]](Left(_))))
     .serverLogic(principal => input => myBusinessLogic(input, principal).andThen(_.map(_.left.map[Either[ErrorInfo, EligibilityError]](Right(_))))
```

It would be nice to have an experience like this: 

```scala
endpoint
      .errorOut(statusCode(StatusCode.NotFound))
      .securityErrorOut(
       oneOf[SecurityError](
          oneOfVariant(StatusCode.Unauthorized, emptyOutputAs(Unauthorized)),
          oneOfVariant(StatusCode.Forbidden, emptyOutputAs(Forbidden))
        )
```

Which would not impact what happens in the logics:

```scala
   myEndpoint
     .serverSecurityLogic(SecurityPolicy.doStuff) // return Either[SecurityError, Principal]
     .serverLogic(principal => input => myBusinessLogic(input, principal)) // return Either[BusinessError, Output] 
```

Unless there is another approach?

**Comment by adamw** [comment]
You're right that it would be nice to have a separate error channel for security errors. However, this would mean that the `Endpoint` type would gain yet another type parameter (it already has 5!). It took me a long time to come to terms with adding the 5th (security input), yet alone the 6th. However, that's something to consider for tapir v2. 

As for what you can do now, there are the `.errorOutVariant` methods, as well as `.errorOutEither` using which error outputs might be extended, after the security logic is provided. There's also `.mapErrorOut` which might help in flattening the nested eithers. 

But maybe for tapir v2 a dedicated security error output is the way to go
